### PR TITLE
Use `infinite-list` to eliminate magic numbers

### PR DESCRIPTION
**Pull request by DigitalBrains1**
In primitive templates, we often refer to inputs by their index in the list of all inputs. This is cumbersome and error-prone, and makes maintenance a headache. GHC 9.2 is very quick on reporting an `incomplete-uni-patterns` warning, but by using a provably infinite list, we can graft an arbitrary number of binders from that list without warnings and name our indices with them.

All occurences of these magic numbers that I could find have been replaced by named indices, except for where this is impossible (notably `clash-lib/prims` which contains many, many more).

The `TemplateFunction`s in `clash-lib:Clash.Primitives.Intel.ClockGen` were technically incorrect. It didn't matter though: the `used` field is not actually currently used in Clash, and if the `valid` field is too optimistic, the render function (last field) will still throw an error message to the user. This different error message provides more information in several ways, but it crucially lacks the location in the user code that triggered the error, which is a significant disadvantage.
The problems were these:
  - `used` for `altpll` was overly broad, listing all arguments. Collectively, the two `TemplateFunction`s did need all arguments, though.
  - `used` for `alteraPll` lists 21 inputs but the function has only 6. Most likely the up-to-16-tuple that is one of the constraints was accidentally thought to be expanded into separate constraints.
  - `valid` for `alteraPllTemplate` doesn't check `bbResults` for the correct shape.
  - The `valid`s for the `QsysTF`s don't check the parsed constraint arguments for validity.
  - Conversely, they do  check the `name` argument and the results for validity, but these are only used by `altpllTF` and `alteraPllTF` and are also already checked there.

The `used` fields were fixed. The `valid` functions have superfluous checks removed, but no checks were added. The net result is that the same error messages are emitted as before this patch.

## Still TODO:

  - ~~Write a changelog entry (see changelog/README.md)~~
  - [x] Check copyright notices are up to date in edited files

### CODE REVIEW

**Comment by DigitalBrains1** [comment]
> The `valid` functions have superfluous checks removed, but no checks were added as that would make the error message worse.

Alternatively, I could just replace all `valid` functions by `const True`, meaning we get the more useful error message for all invalid inputs. I think that's a good solution too, but I didn't want to change the code more than necessary (for a certain value of necessary).

Note you can trigger the not-pretty `valid` error message currently by using Unicode characters in the name of the PLL (this is a bug we should fix, but it's nice to abuse here):
```
IntelXilinxClock.hs:16:1: error:
    
    Clash.Netlist.BlackBox(216): Couldn't instantiate blackbox for Clash.Intel.ClockGen.alteraPll. Verification procedure reported:
    
    Template function for returned False
    
    The source location of the error is not exact, only indicative, as it is acquired 
    after optimizations. The actual location of the error can be in a function that is 
    inlined. To prevent inlining of those functions, annotate them with a NOINLINE pragma.
   |
16 | f = alteraPll (SSymbol @"Je suis désolé")
   | ^
```

**Review by martijnbastiaan — APPROVED** [review]
Supersedes https://github.com/clash-lang/clash-compiler/pull/2435?

File: clash-lib/src/Clash/Primitives/Intel/ClockGen.hs [hunk 1]
```diff
@@ -253,15 +283,19 @@ altpllQsysTemplate bbCtx = case bbInputs bbCtx of
         |]
       in
         pure bbText
-  inps -> error ("altpllQsysTemplate: bad bInputs: " <> show inps)
+  | otherwise
+  = error ("altpllQsysTemplate: bad bbContext: " <> show bbCtx)
```
[the remark this answers is not in the document]
Comment (DigitalBrains1) [reply clash-lib/src/Clash/Primitives/Intel/ClockGen.hs:287]: Funny you mention it, that is literally what I did to look at that black box context with the are-there-6-or-21-inputs conundrum.

**Comment by christiaanb** [comment]
@DigitalBrains1 So do we still get location info in user code or not now that you've changed `valid` to `const true`? If we no longer get the location in user code I think I'd rather keep the old `valid` code. Then in a follow-up PR instead of having `valid` return a `Bool`, we could have it return:
````
data Success = Success | Failure String
````
instead, so that we can return more information than just `False` in case the blackbox input is ill-formed.

**Comment by DigitalBrains1** [comment]
I only removed checks that were superfluous.  Both `altpllTF` and `altpllQsysTF` where checking the arguments only `altpllTF` uses. But combined they still perform the same checks on the primitive. Same for `alteraPll`. I should have mentioned this explicitly in the cover letter.

I think it's better to change `TemplateFunction` from
```haskell
  TemplateFunction
    :: [Int]
    -- ^ Used arguments
    -> (BlackBoxContext -> Bool)
    -- ^ Validation function. Should return 'False' if function can't render
    -- given a certain context.
    -> (forall s . Backend s => BlackBoxContext -> Lazy.State s Doc)
    -- ^ Render function
    -> TemplateFunction
```
to
```haskell
  TemplateFunction
    :: [Int]
    -- ^ Used arguments
    -> (forall s . Backend s => BlackBoxContext -> Lazy.State s (Either String Doc))
    -- ^ Render function
    -> TemplateFunction
```
or something like that instead of improving the `valid` function.

**Comment by DigitalBrains1** [comment]
@mergify rebase

**Pull request merged**